### DISCUSSION TITLE
Add custom location

### DISCUSSION THREAD

**Discussion by etfz** [nginx-proxy/nginx-proxy#2202] [Q&A]
Hi,

I have read the readme, but I am still a little confused as to how custom locations are configured. I need to include the following location block for a `VIRTUAL_HOST=example.com`:
```
location /notifications/hub {
  proxy_pass http://example.com:3012;
}
```
Am I right in thinking the configuration for overriding a location block would be the same as for adding a new block? I tried putting the above in a file named `example.com_559daddb51fd3028052fd33876b54271979ef4d9_location_override` in my bound `vhost.d` directory, and I have confirmed that the file is present at `/etc/nginx/vhost.d` inside the nginx-proxy container, but the block is nowhere to be seen in `nginx -T`. I see no errors in the container log.

**Comment by SchoNie** [comment]
Do you also have `VIRTUAL_PATH=/notifications/hub` environment set on the container?

**Reply by etfz** [reply]
I do not. The container is served on http://example.com. The location configuration is just for one specific path of the service.

**Reply by SchoNie** [reply]
Ok, but the hash is generated from `VIRTUAL_PATH` and because it's not set [the file is not included.](https://github.com/nginx-proxy/nginx-proxy/blob/main/nginx.tmpl#L172)

You can still include for the / path: `For convenience, the _${PATH_HASH} part can be omitted if the path is /:`
Then the file should be renamed to: `example.com_location_override`
Be carefull going this route: 
> When an override file exists, the location block that is normally created by nginx-proxy is not generated. Instead, the override file is included via the [nginx include directive](https://nginx.org/en/docs/ngx_core_module.html#include).
You are responsible for providing a suitable location block in your override file as required for your service

So then you should also add the location block for / in the override file:
```
location / {
  proxy_pass http://example.com;
}
location /notifications/hub {
  proxy_pass http://example.com:3012;
}
```
I think you also need to proxy_pass example.com on ip or cname. Because `example.com` is already defined as an upstream block.

**Reply by etfz** [reply]
Alright. So it is not possible to _add_ a location block for an arbitrary subpath without needing to rebuild the root location block?

**Reply by etfz** [reply]
Actually, could I not just use the per virtual host configuration?
https://github.com/nginx-proxy/nginx-proxy#per-virtual_host

Edit: Seems like you are right about the upstream issue.

**Reply by SchoNie** [reply]
Yes the per virtual host configuration would be better for you. Then the / example.com block is created. And your custom block is included.
Renaming to `example.com_location` is enough (remove _override)
Just have to deal with the upstream name. Which you can try to [set the SHA1_UPSTREAM_NAME environment variable to true](https://github.com/nginx-proxy/nginx-proxy#unhashed-vs-sha1-upstream-names)

[Multiport support is discussed in a few posts](https://github.com/nginx-proxy/nginx-proxy/pull/2130)

**Reply by etfz** [reply]
I tried the `SHA1_UPSTREAM_NAME` setting, but then I get the following error:
```
host not found in upstream "example.com" in /etc/nginx/vhost.d/example.com:2
```
I don't normally use upstream blocks in nginx. Not sure what's the issue here.

Using a `_location` file would create it as a nested location block, right? Does that work?

**Reply by SchoNie** [reply]
If you could post some code examples, that would help. 
But I tried to reproduce your scenario:

**docker-compose.yml**
```yaml
version: '2'

services:
  nginx-proxy:
    image: nginxproxy/nginx-proxy
    ports:
      - "80:80"
    volumes:
      - /var/run/docker.sock:/tmp/docker.sock:ro
      - ./vhost.d:/etc/nginx/vhost.d
    environment:
      - SHA1_UPSTREAM_NAME=true

  whoami:
    image: jwilder/whoami
    expose:
      - "8000"
    environment:
      - VIRTUAL_HOST=sub.example.com
      - VIRTUAL_PORT=8000
```

**vhost.d/sub.example.com_location:**
```
location /notifications/hub {
  proxy_pass https://pl.lipsum.com/;
}
```

`curl -H "Host: sub.example.com" http://localhost/` gives me the whoami output.
`curl -H "Host: sub.example.com" http://localhost/notifications/hub/` I see proxy is forwarding to http://example.com:3012 (but in my case there is no service listening).

**Reply by etfz** [reply]
Ok, I have to confess that example.com is not the exact domain I've been using. It seems like things break when you use subdomains. So if you take your configuration and use `sub.example.com` instead.

**Reply by SchoNie** [reply]
Yes, I can reproduce with. `sub.example.com` but that `host not found in upstream` error seems to be correct.
Nginx uses `resolver 127.0.0.11;` 127.0.0.11 is the DNS service inside Docker, which can resolve the service name to an ipv4 address.

`sub.example.com` is a non-existent domain so `host not found in upstream`
But if I use a correct existing subdomain, it's working as expected.

I edited [my previous comment](https://github.com/nginx-proxy/nginx-proxy/discussions/2202#discussioncomment-5406134) to a subdomain scenario with an existing subdomain I just tested.

**Reply by etfz** [reply]
Alright. Apparently I just had a number of incorrect assumptions. I did have an issue of DNS, but I thought that nginx would just respond with "service unavailable" or similar for those requests, rather than refusing to start.

In the end, what I had to do was to add the proxied container to a Docker network together with nginx-proxy (Docker container name DNS lookup does not work in the default network, apparently), and then just use the container name as the proxy target.
```
location = /notifications/hub {
        proxy_pass http://vaultwarden:3012;
}
```
